Make 'tools/wakeuptime' use stack maps

sandip4n opened the pull request:
This modifies `tools/wakeuptime` to use the stack maps feature. Rather than using manual, depth-limited stack walking with some architecture-specific code, using stack maps makes the script usable across different architectures. This also adds command line arguments to let the user to set the minimum and maximum amount of time in microseconds over which traces are stored.

yonghong-song commented:
[buildbot, ok to test]

yonghong-song commented on tools/wakeuptime.py:
```diff
@@ -195,30 +180,43 @@ def signal_ignore(signal, frame):
+        # check for an ENOMEM error
+        if k.w_k_stack_id == -errno.ENOMEM:
+            missing_stacks += 1
+            continue
```
Another missing stacks will be the one being silently replaced ones and there are no ways to counting them right now. So I guess we can ignore them and your counting with -ENOMEM is the only option we can do now.

goldshtn commented on tools/wakeuptime.py:
```diff
@@ -118,41 +122,16 @@ def signal_ignore(signal, frame):
     delta = bpf_ktime_get_ns() - *tsp;
     delta = delta / 1000;
-    if (delta < MINBLOCK_US)
+    if ((delta < MINBLOCK_US) || (delta > MAXBLOCK_US))
```
Filter out threads that sleep for a long interval for a reason that isn’t interesting, like accept().

sandip4n commented on tools/wakeuptime.py:
```diff
@@ -118,41 +122,16 @@ def signal_ignore(signal, frame):
     // calculate delta time
```
Ah, I missed that. I have added a new commit with this functionality.

goldshtn commented:
I think as long as we keep the old version in tools/old, it’s ok to update the man page to require Linux 4.6 (which is where stack maps were introduced IIRC).

brendangregg commented:
Thanks @sandip4n ! Tested, works.

As @goldshtn said, I'd put the old version in tools/old, and in the man page under "REQUIREMENTS" I'd write like: "Requires BPF_MAP_TYPE_STACK_TRACE, which was added to Linux 4.6.". Some people have this backported to older kernel versions, so I think that sentence covers all bases: lets people know the actual thing it needs, and says which kernel version it was added to, but without insisting on a kernel version.

sandip4n commented:
@yonghong-song @goldshtn @brendangregg Thanks for your inputs.
As requested, I have updated the manpage and examples and moved the original `wakeuptime` script under `tools/old`.

brendangregg commented:
Great, thanks!
